include/micm/util/cuda_sparse_matrix.hpp

```diff
@@ -12,82 +18,104 @@ namespace micm
+      this->param_ = other.param_;
+      this->param_.d_data_ = nullptr;
+      micm::cuda::CHECK_CUDA_ERROR(micm::cuda::MallocVector(this->param_, this->data_.size()), __FILE__, __LINE__, "cudaMalloc");
+      micm::cuda::CHECK_CUDA_ERROR(micm::cuda::CopyToDeviceFromDevice(this->param_, other.param_), __FILE__, __LINE__, "cudaMemcpyDeviceToDevice");
     }
 
     CudaSparseMatrix(const CudaSparseMatrix& other)
         : SparseMatrix<T, OrderingPolicy>(other)
     {
+      this->param_.d_data_ = nullptr;
     }
 
     CudaSparseMatrix(CudaSparseMatrix&& other) noexcept
-        : SparseMatrix<T, OrderingPolicy>(SparseMatrix<T, OrderingPolicy>::create(other.number_of_blocks_))
+        : SparseMatrix<T, OrderingPolicy>(other)
     {
-      this->data_ = std::move(other.data_);
-      this->param_ = std::move(other.param_);
+      this->param_.d_data_ = nullptr;
+      std::swap(this->param_, other.param_);
     }
 
     CudaSparseMatrix& operator=(const CudaSparseMatrix& other)
     {
-      return *this = CudaSparseMatrix(other);
+      SparseMatrix<T, OrderingPolicy>::operator=(other);
+      this->param_ = other.param_;
+      this->param_.d_data_ = nullptr;
+      micm::cuda::CHECK_CUDA_ERROR(micm::cuda::MallocVector(this->param_, this->data_.size()), __FILE__, __LINE__, "cudaMalloc");
+      micm::cuda::CHECK_CUDA_ERROR(micm::cuda::CopyToDeviceFromDevice(this->param_, other.param_), __FILE__, __LINE__, "cudaMemcpyDeviceToDevice");
+      return *this;
     }
 
     CudaSparseMatrix& operator=(CudaSparseMatrix&& other) noexcept
     {
-      std::swap(this->data_, other.data_);
-      std::swap(this->param_, other.param_);
-      std::swap(this->number_of_blocks_, other.number_of_blocks_);
-      std::swap(this->row_ids_, other.row_ids_);
-      std::swap(this->row_start_, other.row_start_);
+      if (this != &other)
+      {
+        SparseMatrix<T, OrderingPolicy>::operator=(std::move(other));
+        std::swap(this->param_, other.param_);
+      }
+      return *this;
     }
 
     ~CudaSparseMatrix() requires(std::is_same_v<T, double>)
     {
-      micm::cuda::FreeVector(param_);
+      std::cout << "Freeing device memory at address: " << this->param_.d_data_ << std::endl;
+      micm::cuda::CHECK_CUDA_ERROR(micm::cuda::FreeVector(this->param_), __FILE__, __LINE__, "cudaFree");
+      this->param_.d_data_ = nullptr;
     }
 
     ~CudaSparseMatrix()
     {
+      this->param_.d_data_ = nullptr;
     }
 
-    int CopyToDevice()
+    void CopyToDevice()
     {
-      return micm::cuda::CopyToDevice(param_, this->data_);
+       micm::cuda::CHECK_CUDA_ERROR(micm::cuda::CopyToDevice(this->param_, this->data_), __FILE__, __LINE__, "cudaMemcpyHostToDevice");
     }
-    int CopyToHost()
+
+    void CopyToHost()
     {
-      return micm::cuda::CopyToHost(param_, this->data_);
+      micm::cuda::CHECK_CUDA_ERROR(micm::cuda::CopyToHost(this->param_, this->data_), __FILE__, __LINE__, "cudaMemcpyDeviceToHost");
```

Comment:
Is it worth having default `CopyToHost()` and `CopyToDevice()` functions that do nothing and then make these `requires(std::is_same_v<T, double>)`?

Reply:
Add the `static_assert` instead for consistency.